templates/vpc.yml

```diff
@@ -1,88 +1,472 @@
       Tags:
         - Key: Name
-          Value: { Ref: 'AWS::StackName' }
+          Value: !Ref AWS::StackName
 
-  Gateway:
+  InternetGateway:
     Type: AWS::EC2::InternetGateway
     Condition: CreateVpcResources
     Properties:
       Tags:
         - Key: Name
-          Value: { Ref: 'AWS::StackName' }
+          Value: !Ref AWS::StackName
 
-  GatewayAttachment:
+  InternetGatewayAttachment:
     Type: AWS::EC2::VPCGatewayAttachment
     Condition: CreateVpcResources
     Properties:
-      InternetGatewayId: { Ref: Gateway }
-      VpcId: { Ref: Vpc }
+      InternetGatewayId: !Ref InternetGateway
+      VpcId: !Ref Vpc
 
-  Subnet0:
+  PublicSubnet1:
     Type: AWS::EC2::Subnet
-    Condition: CreateVpcResources
+    Condition: HasPublicSubnet1
     Properties:
-      AvailabilityZone:
-        "Fn::If":
-          - "UseSpecifiedAvailabilityZones"
-          - { "Fn::Select": [ 0, { Ref: AvailabilityZones } ] }
-          - { "Fn::Select": [ 0, { "Fn::GetAZs": '' } ] }
-      CidrBlock: 10.0.1.0/24
-      VpcId: { Ref: Vpc }
+      VpcId: !Ref Vpc
+      AvailabilityZone: !If [
+        UseSpecifiedAvailabilityZones,
+        !Select [ 0, !Ref AvailabilityZones ],
+        !Select [ 0, !GetAZs "" ]
+      ]
+      CidrBlock: 10.0.0.0/19
+      MapPublicIpOnLaunch: true
       Tags:
         - Key: Name
-          Value: { Ref: 'AWS::StackName' }
+          Value: !Sub ${AWS::StackName} Public Subnet (AZ1)
 
-  Subnet1:
+  PublicSubnet2:
     Type: AWS::EC2::Subnet
-    Condition: CreateVpcResources
+    Condition: HasPublicSubnet2
+    Properties:
+      VpcId: !Ref Vpc
+      AvailabilityZone: !If [
+        UseSpecifiedAvailabilityZones,
+        !Select [ 1, !Ref AvailabilityZones ],
+        !Select [ 1, !GetAZs "" ]
+      ]
+      CidrBlock: 10.0.32.0/19
+      MapPublicIpOnLaunch: true
+      Tags:
+        - Key: Name
+          Value: !Sub ${AWS::StackName} Public Subnet (AZ2)
+
+  PublicSubnet3:
+    Type: AWS::EC2::Subnet
+    Condition: HasPublicSubnet3
     Properties:
-      AvailabilityZone:
-        "Fn::If":
-          - "UseSpecifiedAvailabilityZones"
-          - { "Fn::Select": [ 1, { Ref: AvailabilityZones } ] }
-          - { "Fn::Select": [ 1, { "Fn::GetAZs": '' } ] }
-      CidrBlock: 10.0.2.0/24
-      VpcId: { Ref: Vpc }
+      VpcId: !Ref Vpc
+      AvailabilityZone: !If [
+        UseSpecifiedAvailabilityZones,
+        !Select [ 2, !Ref AvailabilityZones ],
+        !Select [ 2, !GetAZs "" ]
+      ]
+      CidrBlock: 10.0.128.0/19
+      MapPublicIpOnLaunch: true
       Tags:
         - Key: Name
-          Value: { Ref: 'AWS::StackName' }
+          Value: !Sub ${AWS::StackName} Public Subnet (AZ3)
 
-  Routes:
+  PublicSubnet4:
+    Type: AWS::EC2::Subnet
+    Condition: HasPublicSubnet4
+    Properties:
+      VpcId: !Ref Vpc
+      AvailabilityZone: !If [
+        UseSpecifiedAvailabilityZones,
+        !Select [ 3, !Ref AvailabilityZones ],
+        !Select [ 3, !GetAZs "" ]
```

Comment:
Internally we discourage the use of GetAZ call. Previously us-east had an AZ removed (it was C for us, but likely different for others). Because it was in the middle of the list, the calls to GetAZ changed and wanted to re-create our VPC, Redis instances etc.

Reply:
Nope!

Reply:
In our case, you can at least provide the AZ's if you need to. I'm inclined to keep supporting an empty value that causes this lookup unless there is a better option. I'll do some investigating.

Reply:
![20171220-0rqont2ed18sny3](https://user-images.githubusercontent.com/15758/34200317-c570ddd0-e5c4-11e7-9aa2-c10a57a6d038.gif)

With a decent description it would probably be workable.

Reply:
The difficulty of this approach is that it's not possible to figure out a length of a list, or to check if an offset exists, so we'd have to also have an `AvailabilityZonesCount` or similar param. Realistically, we probably need to have this anyway, otherwise folks will get errors if they provide less AZ's than their region should theoretically allow.

Reply:
How about this?